[InstCombine] Remove some of the complexity-based canonicalization

nikic opened the pull request:
The idea behind is that the canonicalization allows us to handle less patterns, because we know that some will be canonicalized away. This is indeed very useful to e.g. know that constants are always on the right.

However, this is only useful if the canonicalization is actually reliable. This is the case for constants, but not for arguments: Moving these to the right makes it look like the "more complex" expression is guaranteed to be on the left, but this is not actually the case in practice. It fails as soon as you replace the argument with another instruction.

The end result is that it looks like things correctly work in tests, while they actually don't. We use the "thwart complexity-based canonicalization" trick to handle this in tests, but it's often a challenge for new contributors to get this right, and based on the regressions this PR originally exposed, we clearly don't get this right in many cases.

For this reason, I think that it's better to remove this complexity canonicalization. It will make it much easier to write tests for commuted cases and make sure that they are handled.

nikic committed:
Remove some of the complexity-based canonicalization

The idea behind is that the canonicalization allows us to handle
less pattern, because we know that some will be canonicalized away.
This is indeed very useful to e.g. know that constants are always
on the right.

However, the fact that arguments are also canonicalized to the
right seems like it may be doing more damage than good: This means
that writing tests to cover both commuted forms requires special
care ("thwart complexity-based canonicalization").

I think we should consider dropping this canonicalization to make
testing simpler.

nikic commented:
cc @dtcxzyw @goldsteinn to get some early feedback.

dtcxzyw commented:
It is surprising that this patch slightly improves the performance on MultiSource/Benchmarks/McCat/05-eks/eks...

goldsteinn commented on llvm/test/Transforms/PhaseOrdering/AArch64/peel-multiple-unreachable-exits-for-vectorization.ll:
```diff
@@ -9,96 +9,47 @@
 ; CHECK:       exit:
-; CHECK-NEXT:    [[SUM_NEXT_LCSSA:%.*]] = phi i64 [ [[SUM_NEXT_PEEL]], [[AT_WITH_INT_CONVERSION_EXIT11_PEEL:%.*]] ], [ [[SUM_NEXT]], [[AT_WITH_INT_CONVERSION_EXIT11]] ]
-; CHECK-NEXT:    ret i64 [[SUM_NEXT_LCSSA]]
+; CHECK-NEXT:    ret i64 [[SUM_NEXT]]
```
Looks like it breaks vectorization here?

nikic replied:
This seems to be a difference in LICM behavior depending on how you write the comparison :/ https://llvm.godbolt.org/z/b6TWfbMGc

nikic replied:
The LICM difference is due to the code in https://github.com/llvm/llvm-project/blob/f0a681640e012356974024dd2971d74fc18f5b48/llvm/lib/Analysis/MustExecute.cpp#L117 I have a fix for it, but need to figure out what's up with this PhaseOrdering test before landing it...

nikic replied:
I think the core problem here is that SCEV cannot compute an exit count for something like this:
```llvm
define void @test(i64 %M, i64 %N) {
entry:
  br label %loop

loop:
  %iv = phi i64 [ 0, %entry ], [ %iv.next, %latch ]
  %cmp1 = icmp ule i64 %iv, %M
  br i1 %cmp1, label %latch, label %error

error:
  call void @error()
  unreachable

latch:
  %iv.next = add nuw i64 %iv, 1
  %exitcond.not = icmp eq i64 %iv, %N
  br i1 %exitcond.not, label %exit, label %loop

exit:
  ret void
}

declare void @error()
```

We know that this is not an infinite loop (thanks to the `add nuw` even if we only look at the first exit). However, the `%loop` exit count would be `%M + 1`, which may overflow. We know that *if this exit is taken*, then that overflow can't happen. But if another exit is taken, it can. Computing the BECount of the whole loop as `umin(%M + 1, %N)` would not be correct. I guess we could return the exit count as `zext(%M) + 1` though, which is what we sometimes do when converting from BECount to trip count.

@fhahn It seems like the whole "peel to make load dereferenceable" approach from https://github.com/llvm/llvm-project/commit/cd0ba9dc58c5806f4e3cc9635ab1f64af6973a83 basically just fixed things by accident. Peeling was never necessary in the first place to make the load dereferenceable, we just failed to see it due to a simple implementation bug. What the peeling actually did is make the exit count of the loop computable...

nikic replied:
I've put up https://github.com/llvm/llvm-project/pull/92206 for the SCEV issue.

nikic replied:
This is fixed now.

goldsteinn commented on llvm/test/Transforms/PGOProfile/chr.ll:
```diff
@@ -1930,15 +1930,15 @@ bb4:
-; CHECK-NEXT:    [[TMP1:%.*]] = and i1 [[TMP0]], [[CMP3]]
-; CHECK-NEXT:    [[TMP2:%.*]] = and i1 [[TMP1]], [[CMP_I]]
-; CHECK-NEXT:    br i1 [[TMP2]], label [[BB1:%.*]], label [[ENTRY_SPLIT_NONCHR:%.*]], !prof [[PROF15]]
+; CHECK-NEXT:    [[TMP1:%.*]] = freeze i1 [[CMP3]]
```
This looks a bit suspect.

goldsteinn commented on llvm/test/Transforms/InstCombine/add.ll:
```diff
@@ -3286,8 +3286,11 @@ define i32 @add_reduce_sqr_sum_flipped(i32 %a, i32 %b) {
+; CHECK-NEXT:    [[TWO_A:%.*]] = shl i32 [[A]], 1
+; CHECK-NEXT:    [[TWO_A_PLUS_B:%.*]] = add i32 [[TWO_A]], [[B:%.*]]
+; CHECK-NEXT:    [[MUL:%.*]] = mul i32 [[B]], [[TWO_A_PLUS_B]]
+; CHECK-NEXT:    [[ADD:%.*]] = add i32 [[MUL]], [[A_SQ]]
```
regression + below

nikic replied:
Fixed in https://github.com/llvm/llvm-project/commit/0d335f78e45341db53d9f956adcebbb2d2616c9a.

goldsteinn commented on llvm/test/Transforms/InstCombine/bit-checks.ll:
```diff
@@ -808,10 +808,12 @@ define i32 @main7a_logical(i32 %argc, i32 %argc2, i32 %argc3) {
+; CHECK-NEXT:    [[AND2:%.*]] = and i32 [[ARGC]], [[ARGC3:%.*]]
+; CHECK-NEXT:    [[TOBOOL3:%.*]] = icmp ne i32 [[ARGC3]], [[AND2]]
+; CHECK-NEXT:    [[AND_COND_NOT:%.*]] = or i1 [[TOBOOL]], [[TOBOOL3]]
+; CHECK-NEXT:    [[STOREMERGE:%.*]] = zext i1 [[AND_COND_NOT]] to i32
```
regression + below

nikic replied:
Put up https://github.com/llvm/llvm-project/pull/104941 to fix this.

goldsteinn commented on llvm/test/Transforms/InstCombine/fadd.ll:
```diff
@@ -648,8 +648,11 @@ define float @fadd_reduce_sqr_sum_varA_order2(float %a, float %b) {
+; CHECK-NEXT:    [[TWO_A:%.*]] = fmul float [[A]], 2.000000e+00
+; CHECK-NEXT:    [[TWO_A_PLUS_B:%.*]] = fadd float [[TWO_A]], [[B:%.*]]
+; CHECK-NEXT:    [[MUL:%.*]] = fmul float [[B]], [[TWO_A_PLUS_B]]
+; CHECK-NEXT:    [[ADD:%.*]] = fadd reassoc nsz float [[MUL]], [[A_SQ]]
```
regression + below

nikic replied:
I think this one got fixed as a side-effect of https://github.com/llvm/llvm-project/commit/0d335f78e45341db53d9f956adcebbb2d2616c9a as well.

goldsteinn commented on llvm/test/Transforms/InstCombine/fast-basictest.ll:
```diff
@@ -549,8 +549,9 @@ define float @test16_reassoc(float %a, float %b, float %z) {
-; CHECK-NEXT:    [[F:%.*]] = fmul fast float [[TMP1]], [[A:%.*]]
+; CHECK-NEXT:    [[C:%.*]] = fmul fast float [[Z:%.*]], -4.000000e+01
+; CHECK-NEXT:    [[TMP1:%.*]] = fneg fast float [[A:%.*]]
+; CHECK-NEXT:    [[F:%.*]] = fmul fast float [[C]], [[TMP1]]
```
regression + below

nikic replied:
I think this is related to this weird transform: https://github.com/llvm/llvm-project/blob/97be79ca126c1a0e174fdbc345a28868edc7cdc7/llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp#L2685 It pushes the fneg into one of the operands, so it's fundamentally asymmetric. I believe for integers we perform the transform the other way around, such that it is symmetric.

goldsteinn commented on llvm/test/Transforms/InstCombine/fast-math.ll:
```diff
@@ -711,9 +711,9 @@ define double @sqrt_intrinsic_three_args5(double %x, double %y) {
-; CHECK-NEXT:    [[SQRT:%.*]] = fmul fast double [[FABS]], [[SQRT1]]
+; CHECK-NEXT:    [[MUL:%.*]] = fmul fast double [[X:%.*]], [[X]]
+; CHECK-NEXT:    [[MUL2:%.*]] = fmul fast double [[Y:%.*]], [[MUL]]
+; CHECK-NEXT:    [[SQRT:%.*]] = call fast double @llvm.sqrt.f64(double [[MUL2]])
```
regression

nikic replied:
Fixed by https://github.com/llvm/llvm-project/commit/60bffe221a1d615ffc7c6b632287d0fbd27ef863.

goldsteinn commented on llvm/test/Transforms/InstCombine/or-xor.ll:
```diff
@@ -261,9 +261,12 @@ define i32 @test10_commuted(i32 %A, i32 %B) {
-; CHECK-NEXT:    ret i32 -1
+; CHECK-NEXT:    [[XOR2:%.*]] = xor i32 [[B]], [[NOT]]
+; CHECK-NEXT:    [[OR:%.*]] = or i32 [[XOR1]], [[XOR2]]
+; CHECK-NEXT:    ret i32 [[OR]]
```
regression + below

nikic replied:
Fixed in https://github.com/llvm/llvm-project/commit/534701d5f93369e822f3afc9670e3a42b08dfc6f.

goldsteinn commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -132,21 +132,18 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+        match(V, m_FNeg(PatternMatch::m_Value())))
+      return 2;
+
+    return 3;
```
Do we also need to change the ordering of `cast/not/neg` vs other instructions?
Also, is the diff cleaner if you keep the original values but just drop the `isa<Argument>(V)` case?

nikic replied:
I'm not sure I get what you mean here. The exact values used here don't matter, it's just a relative order.

goldsteinn commented:
The `freeze` reassos and vectorization change I think deserve a closer look. Otherwise just seems like some misc regressions in `InstCombine` none of which seems like a huge deal (Maybe the `B == (B & A) & D == (D & A)` case is important).

@dtcxzyw am I reading it right that you had no diff in your benchmarks? That seems surprising...

dtcxzyw commented:
> am I reading it right that you had no diff in your benchmarks? That seems surprising...

No. I just canceled the run as I thought it would produce tons of diff :(

goldsteinn commented:
> > am I reading it right that you had no diff in your benchmarks? That seems surprising...
> 
> No. I just canceled the run as I thought it would produce tons of diff :(

I see (another case for: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/355). Can you run it? Ill try and get `-stats` to work locally and post them here.

dtcxzyw commented:
> Can you run it? Ill try and get -stats to work locally and post them here.

I cannot run it as LLVM doesn't support the cost estimation of struct types :(

goldsteinn commented:
> > Can you run it? Ill try and get -stats to work locally and post them here.
> 
> I cannot run it as LLVM doesn't support the cost estimation of struct types :(

Err misunderstanding, I mean can you just generate your normal diffs. Ill do the cost estimation stuff locally and just post the results here.

dtcxzyw commented:
> > > Can you run it? Ill try and get -stats to work locally and post them here.
> > 
> > 
> > I cannot run it as LLVM doesn't support the cost estimation of struct types :(
> 
> Err misunderstanding, I mean can you just generate your normal diffs. Ill do the cost estimation stuff locally and just post the results here.

Done. The IR diff basically looks fine to me.

The main problem is that some `icmp pred A, B` and `icmp swap(pred) B, A` pairs are not CSEed now.
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591786430.

BTW I find that `SimplifyDemandBits` breaks is_pow2 idioms.
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591795230.
I will post a patch later.

There are some missed optimizations which should be handled explicitly.
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591800638
and https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591803409.

goldsteinn commented:
> > > > Can you run it? Ill try and get -stats to work locally and post them here.
> > > 
> > > 
> > > I cannot run it as LLVM doesn't support the cost estimation of struct types :(
> > 
> > 
> > Err misunderstanding, I mean can you just generate your normal diffs. Ill do the cost estimation stuff locally and just post the results here.
> 
> Done. The IR diff basically looks fine to me.
> 
> The main problem is that some `icmp pred A, B` and `icmp swap(pred) B, A` pairs are not CSEed now. See [dtcxzyw/llvm-opt-benchmark#583 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591786430).
> 
Ah, truthfully that and commutative binops imo make a case for keeping this as is.

nikic commented:
> > > > > Can you run it? Ill try and get -stats to work locally and post them here.
> > > > 
> > > > 
> > > > I cannot run it as LLVM doesn't support the cost estimation of struct types :(
> > > 
> > > 
> > > Err misunderstanding, I mean can you just generate your normal diffs. Ill do the cost estimation stuff locally and just post the results here.
> > 
> > 
> > Done. The IR diff basically looks fine to me.
> > The main problem is that some `icmp pred A, B` and `icmp swap(pred) B, A` pairs are not CSEed now. See [dtcxzyw/llvm-opt-benchmark#583 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591786430).
> 
> Ah, truthfully that and commutative binops imo make a case for keeping this as is.

Our main CSE passes (EarlyCSE and GVN) have support for handling commutative operations. I think in this case the failure is probably from a pass like SimplifyCFG where we may fail to hoist/sink "non-identical" instructions.

The problem with relying on InstCombine complexity-based canonicalization for this purpose is that it's very weak. It will handle the case where the icmp has an instruction and argument operand, but if both are instructions, then they will not be canonicalized and we're back to the same problem. So I think to handle this properly we still need explicit commutative operation support.

goldsteinn commented:
> > > > > > Can you run it? Ill try and get -stats to work locally and post them here.
> > > > > 
> > > > > 
> > > > > I cannot run it as LLVM doesn't support the cost estimation of struct types :(
> > > > 
> > > > 
> > > > Err misunderstanding, I mean can you just generate your normal diffs. Ill do the cost estimation stuff locally and just post the results here.
> > > 
> > > 
> > > Done. The IR diff basically looks fine to me.
> > > The main problem is that some `icmp pred A, B` and `icmp swap(pred) B, A` pairs are not CSEed now. See [dtcxzyw/llvm-opt-benchmark#583 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591786430).
> > 
> > 
> > Ah, truthfully that and commutative binops imo make a case for keeping this as is.
> 
> Our main CSE passes (EarlyCSE and GVN) have support for handling commutative operations. I think in this case the failure is probably from a pass like SimplifyCFG where we may fail to hoist/sink "non-identical" instructions.
> 
> The problem with relying on InstCombine complexity-based canonicalization for this purpose is that it's very weak. It will handle the case where the icmp has an instruction and argument operand, but if both are instructions, then they will not be canonicalized and we're back to the same problem. So I think to handle this properly we still need explicit commutative operation support.

What about making the matchers for commutative ops default to the commutative version (thinks like cmp could be included).
We do that in the new `sd_match` API.
Or have the builder do the canonicalization... although that seems tricky.

nikic commented on llvm/test/Transforms/InstCombine/pr14365.ll:
```diff
@@ -31,7 +31,7 @@ define i32 @test1(i32 %a0) {
 ; CHECK-NEXT:    [[TMP1:%.*]] = lshr i32 [[A0:%.*]], 1
 ; CHECK-NEXT:    [[TMP2:%.*]] = and i32 [[TMP1]], 1431655765
-; CHECK-NEXT:    [[TMP3:%.*]] = sub nsw i32 [[A0]], [[TMP2]]
+; CHECK-NEXT:    [[TMP3:%.*]] = sub i32 [[A0]], [[TMP2]]
```
Fixed by https://github.com/llvm/llvm-project/commit/dd9a99f2b634d95072ae49ebcbe5598877de4985.

nikic commented:
> > > > Can you run it? Ill try and get -stats to work locally and post them here.
> > > 
> > > 
> > > I cannot run it as LLVM doesn't support the cost estimation of struct types :(
> > 
> > 
> > Err misunderstanding, I mean can you just generate your normal diffs. Ill do the cost estimation stuff locally and just post the results here.
> 
> Done. The IR diff basically looks fine to me.
> 
> The main problem is that some `icmp pred A, B` and `icmp swap(pred) B, A` pairs are not CSEed now. See [dtcxzyw/llvm-opt-benchmark#583 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/583#discussion_r1591786430).

I've put up https://github.com/llvm/llvm-project/pull/104805 to address this issue.

Another thing we could do is to add complexity canonicalization to GVN instead. GVN can do this in a principled way, by sorting by value number. Not sure this is actually needed though.

nikic committed:
Push fneg into RHS instead of LHS

This is better because we canonicalize fneg to the right, so this
will cancel out two fnegs.

dtcxzyw commented:
> Failed Tests (20):
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1-bfloat.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1sb.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1sh.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1sw.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1ub.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1uh.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_ld1uw.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_st1-bfloat.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_st1.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_st1b.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_st1h.c
  Clang :: CodeGen/aarch64-sve-intrinsics/acle_sve_st1w.c
  Clang :: CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_loads.c
  Clang :: CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_store.c
  Clang :: CodeGen/attr-counted-by.c
  Clang :: CodeGen/fp-reassoc-pragma.cpp
  Clang :: CodeGen/fp-reciprocal-pragma.cpp
  Clang :: CodeGen/ms-mixed-ptr-sizes.c
  Clang :: Headers/wasm.c

dtcxzyw reviewed (approved):
LGTM. Please fix failed clang tests.

nikic committed:
update clang tests
